File: src/services/legacy/bisheng.ts
/**
 * Bisheng 智能体服务
 * 提供 Bisheng 平台的集成功能，包括 API 调用和 iframe 代理
 */

import * as http from 'http';
import * as httpProxy from 'http-proxy';
import { Logger } from '../utils/logger';
import type { BishengConfig, BishengWorkflow, BishengSession } from '../shared/types';

export class BishengService {
  private logger: Logger;
  private config: BishengConfig;
  private proxyServer: http.Server | null = null;
  private proxy: httpProxy | null = null;
  private isProxyRunning: boolean = false;

  constructor(config: BishengConfig, logger: Logger) {
    this.config = config;
    this.logger = logger;
  }

  /**
   * 初始化服务
   */
  async initialize(): Promise<void> {
    this.logger.info('Initializing Bisheng service...', {
      enabled: this.config.enabled,
      baseUrl: this.config.baseUrl,
      mode: this.config.mode,
    });

    if (!this.config.enabled) {
      this.logger.info('Bisheng service is disabled');
      return;
    }

    // 如果是 iframe 模式，启动代理服务器
    if (this.config.mode === 'iframe') {
      await this.startIframeProxy();
    }

    this.logger.info('Bisheng service initialized successfully');
  }

  /**
   * 启动 iframe 代理服务器
   */
  async startIframeProxy(): Promise<void> {
    if (this.isProxyRunning) {
      this.logger.warn('Iframe proxy is already running');
      return;
    }

    try {
      const httpProxyModule = await import('http-proxy');
      this.proxy = httpProxyModule.createProxyServer({});

      // 代理错误处理
      this.proxy.on('error', (err, req, res) => {
        this.logger.error('Proxy error:', err);
        if (res && !res.headersSent) {
          (res as http.ServerResponse).writeHead(500, {
            'Content-Type': 'text/plain',
          });
          (res as http.ServerResponse).end('代理服务器错误');
        }
      });

      // 创建 HTTP 服务器
      this.proxyServer = http.createServer((req, res) => {
        // 移除 X-Frame-Options 响应头
        this.proxy?.once('proxyRes', (proxyRes) => {
          delete proxyRes.headers['x-frame-options'];
          
          // 添加 CORS 头
          proxyRes.headers['access-control-allow-origin'] = '*';
          proxyRes.headers['access-control-allow-methods'] = 'GET, POST, PUT, DELETE, OPTIONS';
          proxyRes.headers['access-control-allow-headers'] = 'Content-Type, Authorization';
          
          // 添加 Content-Security-Policy 允许 iframe 嵌入
          proxyRes.headers['content-security-policy'] = "frame-ancestors 'self' http://localhost:* http://127.0.0.1:*";
        });

        // 代理请求到 Bisheng 前端
        const target = new URL(this.config.frontendUrl);
        this.proxy?.web(req, res, {
          target: `${target.protocol}//${target.host}`,
          changeOrigin: true,
          ws: true,
        });
      });

      // WebSocket 支持
      this.proxyServer.on('upgrade', (req, socket, head) => {
        const target = new URL(this.config.frontendUrl);
        this.proxy?.ws(req, socket, head, {
          target: `${target.protocol}//${target.host}`,
          changeOrigin: true,
        });
      });

      // 启动服务器
      await new Promise<void>((resolve, reject) => {
        this.proxyServer?.listen(this.config.iframeProxyPort, () => {
          this.logger.info(`Iframe proxy server started on port ${this.config.iframeProxyPort}`);
          this.isProxyRunning = true;
          resolve();
        });

        this.proxyServer?.on('error', (err) => {
          this.logger.error('Failed to start iframe proxy server:', err);
          reject(err);
        });
      });
    } catch (error) {
      this.logger.error('Error starting iframe proxy:', error);
      throw error;
    }
  }

  /**
   * 停止 iframe 代理服务器
   */
  async stopIframeProxy(): Promise<void> {
    if (!this.isProxyRunning || !this.proxyServer) {
      return;
    }

    return new Promise<void>((resolve) => {
      this.proxyServer?.close(() => {
        this.logger.info('Iframe proxy server stopped');
        this.isProxyRunning = false;
        this.proxyServer = null;
        this.proxy = null;
        resolve();
      });
    });
  }

  /**
   * 用户登录
   * 注意：当前 Bisheng 服务器登录接口返回 "Decryption failed" 错误
   * 建议使用硬编码的有效 token 或从配置文件读取
   */
  async login(username: string, password: string): Promise<{ token: string; expiry: number }> {
    try {
      this.logger.info('Attempting login', { username });

      const response = await fetch(`${this.config.baseUrl}/api/v1/user/login`, {
        method: 'POST',
        headers: {
          'accept': 'application/json',
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          user_name: username,
          password: password,
        }),
      });

      const data = await response.json();
      this.logger.info('Login response received', {
        status: response.status,
        statusCode: data.status_code,
        statusMessage: data.status_message
      });

      // 检查响应状态
      if (!response.ok || data.status_code !== 200) {
        const errorMsg = data.status_message || `HTTP ${response.status}`;
        this.logger.error('Login failed', { error: errorMsg, data });
        throw new Error(`登录失败: ${errorMsg}`);
      }

      // 从响应中提取 token（Bisheng 标准格式: data.data.access_token）
      const token = data?.data?.access_token;

      if (!token) {
        this.logger.error('No access token in response', { data });
        throw new Error('登录响应中未找到访问令牌');
      }

      // 计算过期时间（默认24小时）
      const expiry = Date.now() + 86400000;

      // 更新配置
      this.config.accessToken = token;
      this.config.tokenExpiry = expiry;

      this.logger.info('Login successful');
      return { token, expiry };
    } catch (error) {
      this.logger.error('Login error:', error);
      throw error;
    }
  }

  /**
   * 获取工作流列表
   */
  async getWorkflows(pageSize: number = 10, pageNum: number = 1): Promise<BishengWorkflow[]> {
    try {
      const url = `${this.config.baseUrl}/api/v1/workflow/list?page_size=${pageSize}&page_num=${pageNum}`;
      
      const response = await fetch(url, {
        method: 'GET',
        headers: {
          'Accept': 'application/json',
          'Authorization': `Bearer ${this.config.accessToken}`,
        },
      });

      if (!response.ok) {
        throw new Error(`Get workflows failed: ${response.status} ${response.statusText}`);
      }

      const result = await response.json();
      
      // 处理不同的响应格式
      let workflows: BishengWorkflow[] = [];
      if (result.data && result.data.data && Array.isArray(result.data.data)) {
        workflows = result.data.data;
      } else if (result.data && result.data.items) {
        workflows = result.data.items;
      } else if (Array.isArray(result.data)) {
        workflows = result.data;
      } else if (Array.isArray(result)) {
        workflows = result;
      }

      this.logger.info(`Retrieved ${workflows.length} workflows`);
      return workflows;
    } catch (error) {
      this.logger.error('Get workflows error:', error);
      throw error;
    }
  }

  /**
   * 调用工作流
   * @param workflowId 工作流 ID
   * @param input 输入数据，格式为 { user_input: string } 或其他字段
   * @param stream 是否使用流式响应
   * @param sessionId 会话 ID（继续对话时必需）
   * @param messageId 消息 ID（继续对话时必需）
   * @param inputNodeId 输入节点 ID（继续对话时必需）
   * @returns ReadableStream 用于读取 SSE 事件流
   */
  async invokeWorkflow(
    workflowId: string,
    input: Record<string, any>,
    stream: boolean = true,
    sessionId?: string,
    messageId?: string,
    inputNodeId?: string
  ): Promise<ReadableStream> {
    try {
      const url = `${this.config.baseUrl}/api/v2/workflow/invoke`;

      // 构建请求体
      const body: any = {
        workflow_id: workflowId,
        stream: stream,
      };

      // 首次调用：不传递 input 参数，让工作流自动启动
      // 继续对话：传递 {node_id: {user_input: "..."}} 格式
      if (sessionId && inputNodeId) {
        // 继续对话
        body.session_id = sessionId;
        body.message_id = messageId ? parseInt(messageId, 10) : 0;
        body.input = {
          [inputNodeId]: input
        };
        this.logger.info('Continuing workflow conversation', {
          workflowId,
          sessionId,
          messageId,
          inputNodeId,
          inputKeys: Object.keys(input)
        });
      } else {
        // 首次调用，不传递 input
        this.logger.info('Starting new workflow', {
          workflowId
        });
      }

      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Accept': 'text/event-stream',
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.config.accessToken}`,
        },
        body: JSON.stringify(body),
      });

      if (!response.ok) {
        const errorText = await response.text();
        this.logger.error('Invoke workflow failed', {
          status: response.status,
          error: errorText
        });
        throw new Error(`工作流调用失败: ${response.status} - ${errorText}`);
      }

      if (!response.body) {
        throw new Error('响应体为空');
      }

      this.logger.info('Workflow invoked successfully, returning stream');
      return response.body;
    } catch (error) {
      this.logger.error('Invoke workflow error:', error);
      throw error;
    }
  }

  /**
   * 停止工作流
   * @param workflowId 工作流 ID
   * @param sessionId 会话 ID
   */
  async stopWorkflow(workflowId: string, sessionId: string): Promise<void> {
    try {
      const url = `${this.config.baseUrl}/api/v2/workflow/stop`;

      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.config.accessToken}`,
        },
        body: JSON.stringify({
          workflow_id: workflowId,
          session_id: sessionId,
        }),
      });

      if (!response.ok) {
        const errorText = await response.text();
        this.logger.error('Stop workflow failed', {
          status: response.status,
          error: errorText
        });
        throw new Error(`停止工作流失败: ${response.status} - ${errorText}`);
      }

      this.logger.info('Workflow stopped successfully', { workflowId, sessionId });
    } catch (error) {
      this.logger.error('Stop workflow error:', error);
      throw error;
    }
  }

  /**
   * 更新配置
   */
  updateConfig(config: Partial<BishengConfig>): void {
    this.config = { ...this.config, ...config };
    this.logger.info('Bisheng config updated');
  }

  /**
   * 获取当前配置
   */
  getConfig(): BishengConfig {
    return { ...this.config };
  }

  /**
   * 检查是否已登录
   */
  isAuthenticated(): boolean {
    if (!this.config.accessToken) {
      return false;
    }

    if (this.config.tokenExpiry && this.config.tokenExpiry < Date.now()) {
      this.logger.warn('Token expired');
      return false;
    }

    return true;
  }

  /**
   * 测试工作流列表 API
   */
  async testWorkflowList(): Promise<{ success: boolean; statusCode: number; data?: any; workflows?: BishengWorkflow[]; error?: string }> {
    try {
      const url = `${this.config.baseUrl}/api/v1/workflow/list?page_size=10&page_num=1`;
      
      this.logger.info('Testing workflow list API:', { url });
      
      const response = await fetch(url, {
        method: 'GET',
        headers: {
          'Accept': 'application/json',
          'Authorization': `Bearer ${this.config.accessToken}`,
        },
      });

      this.logger.info('Workflow list API response status:', response.status);

      const data = await response.json();
      
      if (!response.ok) {
        return {
          success: false,
          statusCode: response.status,
          error: `HTTP ${response.status}: ${response.statusText}`,
          data: data
        };
      }

      // 处理不同的响应格式
      let workflows: BishengWorkflow[] = [];
      if (data.data && data.data.data && Array.isArray(data.data.data)) {
        workflows = data.data.data;
      } else if (data.data && data.data.items) {
        workflows = data.data.items;
      } else if (Array.isArray(data.data)) {
        workflows = data.data;
      } else if (Array.isArray(data)) {
        workflows = data;
      }

      this.logger.info(`Retrieved ${workflows.length} workflows for testing`);
      
      return {
        success: response.ok && (data.status_code === 200 || data.code === 0),
        statusCode: response.status,
        data: data,
        workflows: workflows
      };
    } catch (error) {
      this.logger.error('Test workflow list error:', error);
      return {
        success: false,
        statusCode: 0,
        error: error instanceof Error ? error.message : 'Unknown error'
      };
    }
  }

  /**
   * 测试工作流调用 API
   */
  async testWorkflowInvoke(workflowId: string): Promise<{ success: boolean; statusCode: number; data?: any; isStream?: boolean; error?: string }> {
    try {
      const url = `${this.config.baseUrl}/api/v2/workflow/invoke`;
      
      this.logger.info('Testing workflow invoke API:', { url, workflowId });
      
      const requestBody = {
        workflow_id: workflowId,
        stream: true,
        user_input: {
          query: "你好，这是一个测试消息"
        },
        message_id: 0,
        session_id: "test-session-" + Date.now()
      };

      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.config.accessToken}`,
        },
        body: JSON.stringify(requestBody)
      });

      this.logger.info('Workflow invoke API response status:', response.status);

      if (!response.ok) {
        const errorText = await response.text();
        return {
          success: false,
          statusCode: response.status,
          error: `HTTP ${response.status}: ${errorText}`
        };
      }

      const responseText = await response.text();
      
      // 处理流式响应 (Server-Sent Events)
      if (responseText.startsWith('data: ')) {
        this.logger.info('Processing streaming response');
        const lines = responseText.split('\n');
        let sessionId = '';
        let messageId = '';
        let answer = '';
        
        for (const line of lines) {
          if (line.startsWith('data: ')) {
            try {
              const jsonData = JSON.parse(line.substring(6));
              if (jsonData.session_id) sessionId = jsonData.session_id;
              if (jsonData.data) {
                if (jsonData.data.message_id) messageId = jsonData.data.message_id;
                if (jsonData.data.output_schema && jsonData.data.output_schema.message) {
                  answer += Array.isArray(jsonData.data.output_schema.message) 
                    ? jsonData.data.output_schema.message.join(' ') 
                    : jsonData.data.output_schema.message;
                }
              }
            } catch (e) {
              // 忽略解析错误，继续处理下一行
            }
          }
        }
        
        return {
          success: true,
          statusCode: response.status,
          data: { session_id: sessionId, message_id: messageId, answer: answer },
          isStream: true
        };
      } else {
        // 处理普通 JSON 响应
        try {
          const result = JSON.parse(responseText);
          return {
            success: response.status === 200 && result.code === 0,
            statusCode: response.status,
            data: result,
            isStream: false
          };
        } catch (e) {
          return {
            success: false,
            statusCode: response.status,
            error: 'Failed to parse response as JSON',
            data: responseText
          };
        }
      }
    } catch (error) {
      this.logger.error('Test workflow invoke error:', error);
      return {
        success: false,
        statusCode: 0,
        error: error instanceof Error ? error.message : 'Unknown error'
      };
    }
  }

  /**
   * 运行完整的连接测试
   */
  async runConnectionTests(): Promise<{
    overall: boolean;
    login: { success: boolean; error?: string };
    workflowList: { success: boolean; count: number; error?: string };
    workflowInvoke?: { success: boolean; error?: string };
    details: string[];
  }> {
    const details: string[] = [];
    let loginSuccess = false;
    let workflowListSuccess = false;
    let workflowInvokeSuccess = false;
    let workflowCount = 0;

    try {
      // 1. 测试登录
      details.push('🔍 开始测试 Bisheng API 连接...');
      
      try {
        const loginResult = await this.login(this.config.username, this.config.password);
        if (loginResult && loginResult.token) {
          loginSuccess = true;
          details.push('✅ 登录测试通过');
        } else {
          details.push('❌ 登录失败: 未获取到访问令牌');
          return {
            overall: false,
            login: { success: false, error: '未获取到访问令牌' },
            workflowList: { success: false, count: 0, error: '跳过，登录失败' },
            details
          };
        }
      } catch (error) {
        const errorMsg = error instanceof Error ? error.message : 'Unknown error';
        details.push(`❌ 登录失败: ${errorMsg}`);
        return {
          overall: false,
          login: { success: false, error: errorMsg },
          workflowList: { success: false, count: 0, error: '跳过，登录失败' },
          details
        };
      }

      // 2. 测试工作流列表
      try {
        const workflowResult = await this.testWorkflowList();
        if (workflowResult.success && workflowResult.workflows) {
          workflowListSuccess = true;
          workflowCount = workflowResult.workflows.length;
          details.push(`✅ 工作流列表测试通过，找到 ${workflowCount} 个工作流`);
          
          // 3. 测试工作流调用（如果有工作流）
          if (workflowResult.workflows.length > 0) {
            const firstWorkflow = workflowResult.workflows[0];
            details.push(`🔄 测试工作流调用: ${firstWorkflow.name} (${firstWorkflow.id})`);
            
            try {
              const invokeResult = await this.testWorkflowInvoke(firstWorkflow.id);
              if (invokeResult.success) {
                workflowInvokeSuccess = true;
                details.push('✅ 工作流调用测试通过');
                if (invokeResult.data && invokeResult.data.answer) {
                  details.push(`💬 AI 回复: ${invokeResult.data.answer.substring(0, 100)}...`);
                }
              } else {
                details.push(`❌ 工作流调用失败: ${invokeResult.error}`);
              }
            } catch (error) {
              const errorMsg = error instanceof Error ? error.message : 'Unknown error';
              details.push(`❌ 工作流调用异常: ${errorMsg}`);
            }
          } else {
            details.push('⚠️ 跳过工作流调用测试，没有可用的工作流');
          }
        } else {
          details.push(`❌ 工作流列表测试失败: ${workflowResult.error}`);
        }
      } catch (error) {
        const errorMsg = error instanceof Error ? error.message : 'Unknown error';
        details.push(`❌ 工作流列表测试异常: ${errorMsg}`);
      }

      const overall = loginSuccess && workflowListSuccess;
      
      if (overall) {
        details.push('🎉 所有测试通过！Bisheng API 连接正常。');
      } else {
        details.push('⚠️ 部分测试失败，请检查配置和服务状态。');
      }

      return {
        overall,
        login: { success: loginSuccess },
        workflowList: { success: workflowListSuccess, count: workflowCount },
        workflowInvoke: workflowInvokeSuccess ? { success: true } : undefined,
        details
      };
    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : 'Unknown error';
      details.push(`💥 测试过程发生异常: ${errorMsg}`);
      
      return {
        overall: false,
        login: { success: loginSuccess },
        workflowList: { success: workflowListSuccess, count: workflowCount },
        details
      };
    }
  }

  /**
   * 获取代理状态
   */
  getProxyStatus(): { running: boolean; port: number } {
    return {
      running: this.isProxyRunning,
      port: this.config.iframeProxyPort,
    };
  }

  /**
   * 清理资源
   */
  async cleanup(): Promise<void> {
    this.logger.info('Cleaning up Bisheng service...');
    
    if (this.isProxyRunning) {
      await this.stopIframeProxy();
    }

    this.logger.info('Bisheng service cleanup complete');
  }
}

File: backend-service/app/api/bisheng.py
"""
Bisheng 智能体平台 API 代理
提供与 Bisheng 平台的集成接口
"""
from fastapi import APIRouter, HTTPException, Request
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from typing import Optional, Dict, Any, List
from loguru import logger
import httpx
import json

from app.config import settings

router = APIRouter()


class LoginRequest(BaseModel):
    """登录请求"""
    username: str
    password: str


class WorkflowInvokeRequest(BaseModel):
    """工作流调用请求"""
    workflow_id: str
    input: Dict[str, Any]
    stream: Optional[bool] = True
    session_id: Optional[str] = None
    message_id: Optional[str] = None


class BishengConfigUpdate(BaseModel):
    """Bisheng 配置更新"""
    enabled: Optional[bool] = None
    base_url: Optional[str] = None
    frontend_url: Optional[str] = None
    iframe_proxy_port: Optional[int] = None
    username: Optional[str] = None
    password: Optional[str] = None
    mode: Optional[str] = None
    auto_login: Optional[bool] = None
    save_password: Optional[bool] = None
    timeout: Optional[int] = None
    retry_attempts: Optional[int] = None


@router.post("/login")
async def login(request: LoginRequest):
    """
    登录 Bisheng 平台
    
    Args:
        request: 登录请求，包含用户名和密码
        
    Returns:
        包含 access_token 和过期时间的响应
    """
    if not settings.BISHENG_ENABLED:
        raise HTTPException(status_code=503, detail="Bisheng 服务未启用")
    
    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.post(
                f"{settings.BISHENG_BASE_URL}/api/v1/login",
                json={
                    "username": request.username,
                    "password": request.password
                }
            )
            
            if response.status_code == 200:
                data = response.json()
                logger.info(f"Bisheng 登录成功: {request.username}")
                return {
                    "success": True,
                    "token": data.get("access_token"),
                    "expiry": data.get("expiry", settings.BISHENG_TOKEN_EXPIRY)
                }
            else:
                logger.error(f"Bisheng 登录失败: {response.status_code} - {response.text}")
                raise HTTPException(
                    status_code=response.status_code,
                    detail=f"登录失败: {response.text}"
                )
                
    except httpx.RequestError as e:
        logger.error(f"Bisheng 登录请求错误: {e}")
        raise HTTPException(status_code=503, detail=f"连接 Bisheng 服务失败: {str(e)}")


@router.get("/workflows")
async def get_workflows(
    page_size: Optional[int] = 50,
    page_num: Optional[int] = 1,
    authorization: Optional[str] = None
):
    """
    获取工作流列表
    
    Args:
        page_size: 每页数量
        page_num: 页码
        authorization: Bearer token
        
    Returns:
        工作流列表
    """
    if not settings.BISHENG_ENABLED:
        raise HTTPException(status_code=503, detail="Bisheng 服务未启用")
    
    if not authorization:
        # 尝试使用配置中的 token
        authorization = f"Bearer {settings.BISHENG_ACCESS_TOKEN}" if settings.BISHENG_ACCESS_TOKEN else None
    
    if not authorization:
        raise HTTPException(status_code=401, detail="需要认证")
    
    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.get(
                f"{settings.BISHENG_BASE_URL}/api/v2/workflows",
                params={
                    "page_size": page_size,
                    "page_num": page_num
                },
                headers={
                    "Authorization": authorization
                }
            )
            
            if response.status_code == 200:
                data = response.json()
                logger.info(f"获取工作流列表成功: {len(data.get('data', []))} 个工作流")
                return data
            else:
                logger.error(f"获取工作流列表失败: {response.status_code} - {response.text}")
                raise HTTPException(
                    status_code=response.status_code,
                    detail=f"获取工作流列表失败: {response.text}"
                )
                
    except httpx.RequestError as e:
        logger.error(f"获取工作流列表请求错误: {e}")
        raise HTTPException(status_code=503, detail=f"连接 Bisheng 服务失败: {str(e)}")


@router.post("/workflow/invoke")
async def invoke_workflow(request: WorkflowInvokeRequest, authorization: Optional[str] = None):
    """
    调用工作流
    
    Args:
        request: 工作流调用请求
        authorization: Bearer token
        
    Returns:
        工作流执行结果（流式或非流式）
    """
    if not settings.BISHENG_ENABLED:
        raise HTTPException(status_code=503, detail="Bisheng 服务未启用")
    
    if not authorization:
        authorization = f"Bearer {settings.BISHENG_ACCESS_TOKEN}" if settings.BISHENG_ACCESS_TOKEN else None
    
    if not authorization:
        raise HTTPException(status_code=401, detail="需要认证")
    
    try:
        # 构建请求体
        payload = {
            "input": request.input,
            "stream": request.stream
        }
        
        if request.session_id:
            payload["session_id"] = request.session_id
        if request.message_id:
            payload["message_id"] = request.message_id
        
        async with httpx.AsyncClient(timeout=120.0) as client:
            if request.stream:
                # 流式响应
                async def generate():
                    async with client.stream(
                        "POST",
                        f"{settings.BISHENG_BASE_URL}/api/v2/workflow/invoke/{request.workflow_id}",
                        json=payload,
                        headers={
                            "Authorization": authorization,
                            "Content-Type": "application/json"
                        }
                    ) as response:
                        if response.status_code != 200:
                            error_text = await response.aread()
                            logger.error(f"调用工作流失败: {response.status_code} - {error_text}")
                            yield f"data: {json.dumps({'error': error_text.decode()})}\n\n"
                            return
                        
                        async for chunk in response.aiter_bytes():
                            if chunk:
                                yield chunk
                
                return StreamingResponse(
                    generate(),
                    media_type="text/event-stream",
                    headers={
                        "Cache-Control": "no-cache",
                        "Connection": "keep-alive",
                        "X-Accel-Buffering": "no"
                    }
                )
            else:
                # 非流式响应
                response = await client.post(
                    f"{settings.BISHENG_BASE_URL}/api/v2/workflow/invoke/{request.workflow_id}",
                    json=payload,
                    headers={
                        "Authorization": authorization,
                        "Content-Type": "application/json"
                    }
                )
                
                if response.status_code == 200:
                    logger.info(f"调用工作流成功: {request.workflow_id}")
                    return response.json()
                else:
                    logger.error(f"调用工作流失败: {response.status_code} - {response.text}")
                    raise HTTPException(
                        status_code=response.status_code,
                        detail=f"调用工作流失败: {response.text}"
                    )
                    
    except httpx.RequestError as e:
        logger.error(f"调用工作流请求错误: {e}")
        raise HTTPException(status_code=503, detail=f"连接 Bisheng 服务失败: {str(e)}")


@router.get("/config")
async def get_config():
    """
    获取 Bisheng 配置
    
    Returns:
        当前的 Bisheng 配置
    """
    return {
        "enabled": settings.BISHENG_ENABLED,
        "base_url": settings.BISHENG_BASE_URL,
        "frontend_url": settings.BISHENG_FRONTEND_URL,
        "iframe_proxy_port": settings.BISHENG_IFRAME_PROXY_PORT,
        "username": settings.BISHENG_USERNAME,
        "mode": settings.BISHENG_DEFAULT_MODE,
        "token_expiry": settings.BISHENG_TOKEN_EXPIRY,
        # 不返回密码和 token
    }


@router.post("/config")
async def update_config(config: BishengConfigUpdate):
    """
    更新 Bisheng 配置
    
    Args:
        config: 配置更新请求
        
    Returns:
        更新结果
    """
    try:
        # 更新配置（注意：这里只是示例，实际应该持久化到配置文件或数据库）
        if config.enabled is not None:
            settings.BISHENG_ENABLED = config.enabled
        if config.base_url is not None:
            settings.BISHENG_BASE_URL = config.base_url
        if config.frontend_url is not None:
            settings.BISHENG_FRONTEND_URL = config.frontend_url
        if config.iframe_proxy_port is not None:
            settings.BISHENG_IFRAME_PROXY_PORT = config.iframe_proxy_port
        if config.username is not None:
            settings.BISHENG_USERNAME = config.username
        if config.password is not None:
            settings.BISHENG_PASSWORD = config.password
        if config.mode is not None:
            settings.BISHENG_DEFAULT_MODE = config.mode
        
        logger.info("Bisheng 配置已更新")
        return {"success": True, "message": "配置更新成功"}
        
    except Exception as e:
        logger.error(f"更新 Bisheng 配置失败: {e}")
        raise HTTPException(status_code=500, detail=f"更新配置失败: {str(e)}")


@router.get("/status")
async def get_status():
    """
    获取 Bisheng 服务状态
    
    Returns:
        服务状态信息
    """
    status = {
        "enabled": settings.BISHENG_ENABLED,
        "connected": False,
        "authenticated": False
    }
    
    if not settings.BISHENG_ENABLED:
        return status
    
    try:
        # 检查连接
        async with httpx.AsyncClient(timeout=5.0) as client:
            response = await client.get(f"{settings.BISHENG_BASE_URL}/health")
            status["connected"] = response.status_code == 200
            
        # 检查认证（如果有 token）
        if settings.BISHENG_ACCESS_TOKEN:
            async with httpx.AsyncClient(timeout=5.0) as client:
                response = await client.get(
                    f"{settings.BISHENG_BASE_URL}/api/v2/workflows",
                    params={"page_size": 1, "page_num": 1},
                    headers={"Authorization": f"Bearer {settings.BISHENG_ACCESS_TOKEN}"}
                )
                status["authenticated"] = response.status_code == 200
                
    except Exception as e:
        logger.warning(f"检查 Bisheng 状态失败: {e}")
    
    return status
